vm-validator/src/vm_validator.rs

```diff
@@ -42,11 +52,188 @@ pub trait TransactionValidation: Send + Sync + Clone {
+
+    fn insert_deserialized_module(
+        &self,
+        key: Self::Key,
+        deserialized_code: Self::Deserialized,
+        extension: Arc<Self::Extension>,
+        version: Self::Version,
+    ) -> VMResult<Arc<ModuleCode<Self::Deserialized, Self::Verified, Self::Extension>>> {
+        self.module_cache
+            .insert_deserialized_module(key, deserialized_code, extension, version)
+    }
+
+    fn insert_verified_module(
+        &self,
+        key: Self::Key,
+        verified_code: Self::Verified,
+        extension: Arc<Self::Extension>,
+        version: Self::Version,
+    ) -> VMResult<Arc<ModuleCode<Self::Deserialized, Self::Verified, Self::Extension>>> {
+        self.module_cache
+            .insert_verified_module(key, verified_code, extension, version)
+    }
+
+    fn get_module_or_build_with(
+        &self,
+        key: &Self::Key,
+        builder: &dyn ModuleCodeBuilder<
+            Key = Self::Key,
+            Deserialized = Self::Deserialized,
+            Verified = Self::Verified,
+            Extension = Self::Extension,
+        >,
+    ) -> VMResult<
+        Option<(
+            Arc<ModuleCode<Self::Deserialized, Self::Verified, Self::Extension>>,
+            Self::Version,
+        )>,
+    > {
+        // Get the module that exists in cache.
+        let (module, version) = match self.module_cache.get_module_or_build_with(key, builder)? {
+            None => {
+                return Ok(None);
+            },
+            Some(module_and_version) => module_and_version,
+        };
+
+        // Get the state value that exists in the actual state and compute the hash.
+        let state_value = self
+            .state_view
+            .get_state_value(&StateKey::module_id(key))
+            .map_err(|err| module_storage_error!(key.address(), key.name(), err))?
+            .ok_or_else(|| {
+                PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
+                    .with_message(format!(
+                        "Module {}::{} cannot be found in storage, but exists in cache",
+                        key.address(),
+                        key.name()
+                    ))
+                    .finish(Location::Undefined)
+            })?;
+        let hash = sha3_256(state_value.bytes());
+
+        // If hash is the same - we can use the same module from cache. If the hash is different,
+        // then the state contains a newer version of the code. We deserialize the state value and
+        // replace the old cache entry with the new code.
+        Ok(if module.extension().hash() == &hash {
+            Some((module, version))
+        } else {
+            let compiled_module = self
+                .vm
+                .runtime_environment()
+                .deserialize_into_compiled_module(state_value.bytes())?;
+            let extension = Arc::new(AptosModuleExtension::new(state_value));
+
+            let new_version = version + 1;
```

Comment:
how is this version exactly used? I am asking since we are assigning +1, and need to make sure this makes sense e.g. even if things were published multiple times in between,

Reply:
It is one version per commit, no matter how many things were published in between.